File: Leetcode/C++/BinaryTree/inorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        if(root == NULL)
            return res;
            res.push_back(root->val);

        vector<int> left = inorderTraversal(root->left);
        vector<int> right = inorderTraversal(root->right);

        left.insert(left.end(), res.begin(), res.end());
        left.insert(left.end(), right.begin(), right.end());

        return left;
    }
};

File: Leetcode高频题/inorderTraversal.py
'''
@lsy 2019.11.20
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        
        stack, res = [], []
        node = root
        while stack or node:
            while node:
                s tack.append(node)
                node = node.left
            if stack:
                tmp = stack.pop()
                res.append(tmp.val)
                node = tmp.right
        return res
